lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,647 @@
+            }
+            fs::read_to_string(normalized_path).ok()
+        });
+
+        for module in modules {
+            let normalized_path = normalize_path(&manifest.base_directory_path, &module.source);
+            normalized_path
+                .metadata()
+                .map_err(|_| PublishError::SourceMustBeFile {
+                    module: module.name.clone(),
+                    path: normalized_path.clone(),
+                })?;
+            builder
+                .append_path_with_name(&normalized_path, &module.source)
+                .map_err(|e| {
+                    PublishError::ErrorBuildingPackage(format!("{}", normalized_path.display()), e)
+                })?;
+
+            if let Some(bindings) = &module.bindings {
+                for path in bindings.referenced_files(&manifest.base_directory_path)? {
+                    let normalized_path = normalize_path(&manifest.base_directory_path, &path);
+                    normalized_path
+                        .metadata()
+                        .map_err(|_| PublishError::MissingBindings {
+                            module: module.name.clone(),
+                            path: normalized_path.clone(),
+                        })?;
+                    builder
+                        .append_path_with_name(&normalized_path, &module.source)
+                        .map_err(|e| PublishError::ErrorBuildingPackage(module.name.clone(), e))?;
+                }
+            }
+        }
+
+        // bundle the package filesystem
+        for (_alias, path) in manifest.fs.unwrap_or_default().iter() {
+            let normalized_path = normalize_path(&cwd, path);
+            let path_metadata = normalized_path.metadata().map_err(|_| {
+                PublishError::MissingManifestFsPath(normalized_path.to_string_lossy().to_string())
+            })?;
+            if path_metadata.is_dir() {
+                builder.append_dir_all(path, &normalized_path)
+            } else {
+                return Err(PublishError::PackageFileSystemEntryMustBeDirectory(
+                    path.to_string_lossy().to_string(),
+                )
+                .into());
+            }
+            .map_err(|_| {
+                PublishError::MissingManifestFsPath(normalized_path.to_string_lossy().to_string())
+            })?;
+        }
+
+        builder.finish().ok();
+        let tar_archive_data = builder.into_inner().map_err(|_| PublishError::NoModule)?;
+        let archive_name = "package.tar.gz".to_string();
+        let archive_dir = tempfile::TempDir::new()?;
+        let archive_dir_path: &std::path::Path = archive_dir.as_ref();
+        fs::create_dir(archive_dir_path.join("wapm_package"))?;
+        let archive_path = archive_dir_path.join("wapm_package").join(&archive_name);
+        let mut compressed_archive = fs::File::create(&archive_path).unwrap();
+        let mut gz_enc = GzEncoder::new(&mut compressed_archive, Compression::best());
```

Comment:
Instead of writing to a temporary directory, couldn't we just save to a `Vec<u8>` or write to the final `*.tar.gz` file directly?

Reply:
The file may be very big.

Also, I think that that the tar crate doesn't work on `Vec<u8>` at the moment, but I might be misremembering things.

Reply:
I think we should leave this as is. For now, a tempdir has never caused problems.